src/scheduler/node_info.c

```diff
@@ -5831,12 +6122,80 @@ is_exclhost(place *placespec, enum vnode_sharing sharing)
 	return 0;
 }
 
+/**
+ * @brief	pthread routing to check eligibility for a chunk of nodes
+ *
+ * @param[in,out]	data - th_data_nd_eligible object
+ *
+ * @return int
+ * @retval 0 for success
+ * @retval 1 for error
+ */
+int
+check_node_eligibility_chunk(th_data_nd_eligible *data)
+{
+	int i;
+	int start, end;
+	schd_error *err;
+	resource_resv *resresv;
+	place *pl;
+	char *exclerr_buf;
+	node_info **ninfo_arr;
+
+	if (data == NULL)
+		return 1;
+
+	err = new_schd_error();
+	if (err == NULL)
+		return 1;
+
+	start = data->sidx;
+	end = data->eidx;
+	resresv = data->resresv;
+	pl = data->pl;
+	exclerr_buf = data->exclerr_buf;
+	ninfo_arr = data->ninfo_arr;
+
+	for (i = start; i <= end && ninfo_arr[i] != NULL; i++) {
+		node_info *node;
+
+		node = ninfo_arr[i];
+		if ((!node->nscr.ineligible)) {
+			if (is_vnode_eligible(node, resresv, pl, err) == 0) {
+				node->nscr.ineligible = 1;
+				if (err->status_code != SCHD_UNKWN) {
+					schdlogerr(PBSEVENT_DEBUG3, PBS_EVENTCLASS_NODE, LOG_DEBUG, node->name, NULL, err);
+				}
+				if (node->hostset != NULL) {
+					if ((err->error_code == NODE_NOT_EXCL && is_exclhost(pl, node->sharing))
+							|| sim_exclhost(resresv->server->calendar, resresv, node) == 0) {
+						int j;
+
+						for (j = 0; node->hostset->ninfo_arr[j] != NULL; j++) {
```

Comment:
good catch! I didn't realize that scr.ineligible is being read as well. I think there are a few choices:
1) Don't change anything. I might be wrong about this but from what I understand, even if one thread marks a node src.ineligible while the other is already inside the if ((!node->nscr.ineligible))  block, the other thread will just do some extra work before marking the nscr.ineligible bit. If this is True, then, again I might be wrong about this, but the fact that we send work out in big chunks should help nodes of the same hostset get handled by a single thread. So, this shouldn't happen for many nodes. But please do correct me if my understanding is wrong.
2) Remove the 'if (node->hostset != NULL)' block. This will probably cause some performance degradation, but hopefully not too much as the threads will just have to check each and every node instead of marking all nodes of same hostset as ineligible. Again, I'm not entirely sure if this block can be removed without affecting anything functionally, so please do clarify.
3) If neither of the above can be done, then I guess we'll just have to remove multithreading for this function, as it seems to have iterative dependencies.

Please advise.

Reply:
So you are correct, the block of code will do extra work and mark the scr.ineligible bit.  What I don't know is if there can be a problem between the instructions of the if statement.  The worst case scenario will be all N nodes in the host set will be in N different threads.  We might have N different threads doing the same work and marking the same nodes as ineligible at the same time.  I don't see this as a degradation of performance since before it was iterative as long as you don't see it being a problem.  #2 can't be done.  We used to not have that check, and performance was a dog for systems without hostests (like an ICE).  Things sped up significantly after we added that check.  You likely want to add a comment saying that the different nodes in the hostset might be in different threads.  Someone in the future might come along and do something that is harmful.

Reply:
sorry for the delay in getting back to this @bhroam . So, should I just add a comment there saying that different nodes in hostset could be processed by different threads, in which case there might be some duplication of work?

Reply:
Yeah, that sounds about right.  The main idea I want to portray is that the different nodes in the hostsets might be handled by different threads.  If someone wants to modify other nodes in the hostsets other than marking the ineligible bit, they should be careful.